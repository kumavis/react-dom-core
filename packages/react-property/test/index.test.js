const possibleStandardNamesOriginal = require('../lib/possibleStandardNames');
const {
  getPropertyInfo,
  isCustomAttribute,
  possibleStandardNames
} = require('..');

describe('getPropertyInfo', () => {
  it.each([undefined, null])('returns null for %p', (attribute) => {
    expect(getPropertyInfo(attribute)).toBe(null);
  });

  it('gets info for reserved prop', () => {
    expect(getPropertyInfo('children')).toMatchSnapshot();
  });

  it('gets info for different name attribute', () => {
    expect(getPropertyInfo('acceptCharset')).toMatchSnapshot();
  });

  it('gets info for enumerated HTML attribute', () => {
    expect(getPropertyInfo('contentEditable')).toMatchSnapshot();
  });

  it('gets info for enumerated SVG attribute', () => {
    expect(getPropertyInfo('autoReverse')).toMatchSnapshot();
  });

  it('gets info for HTML boolean attribute', () => {
    expect(getPropertyInfo('allowFullScreen')).toMatchSnapshot();
  });

  it('gets info for DOM property', () => {
    expect(getPropertyInfo('checked')).toMatchSnapshot();
  });

  it('gets for overloaded boolean', () => {
    expect(getPropertyInfo('capture')).toMatchSnapshot();
  });

  it('gets info for HTML attribute that must be a positive number', () => {
    expect(getPropertyInfo('cols')).toMatchSnapshot();
  });

  it('gets info for HTML attribute that must be a number', () => {
    expect(getPropertyInfo('rowSpan')).toMatchSnapshot();
  });

  it('gets info for SVG attribute that need special casing', () => {
    expect(getPropertyInfo('accent-height')).toMatchSnapshot();
  });

  it('gets info for SVG attribute with the xlink namespace', () => {
    expect(getPropertyInfo('xlink:actuate')).toMatchSnapshot();
  });

  it('gets info for SVG attribute with the xml namespace', () => {
    expect(getPropertyInfo('xml:base')).toMatchSnapshot();
  });

  it('gets info for attribute that exists both in HTML and SVG', () => {
    expect(getPropertyInfo('tabIndex')).toMatchSnapshot();
  });

  it('gets info for attributes that accept URLs', () => {
    expect(getPropertyInfo('xlinkHref')).toMatchSnapshot();
    expect(getPropertyInfo('src')).toMatchSnapshot();
  });
});

describe('isCustomAttribute', () => {
  it.each([
    // expected, attribute
    [false, undefined],
    [false, null],
    [false, ''],
    [false, 'dataaria'],
    [false, 'aria'],
    [true, 'aria-'],
    [true, 'aria-live'],
    [false, 'aria-live="polite"'],
    [false, 'data'],
    [true, 'data-'],
    [true, 'data-foo'],
    [false, 'data-foo="bar"']
  ])('returns %j for %j', (expected, attribute) => {
    expect(isCustomAttribute(attribute)).toBe(expected);
  });
});

describe('possibleStandardNames', () => {
  it('matches properties', () => {
    expect(possibleStandardNames).toEqual(possibleStandardNamesOriginal);
  });
});
