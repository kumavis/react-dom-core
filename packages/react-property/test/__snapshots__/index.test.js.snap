// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPropertyInfo gets for overloaded boolean 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": true,
  "attributeName": "capture",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "capture",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 4,
}
`;

exports[`getPropertyInfo gets info for DOM property 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": true,
  "attributeName": "checked",
  "attributeNamespace": null,
  "mustUseProperty": true,
  "propertyName": "checked",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 3,
}
`;

exports[`getPropertyInfo gets info for HTML attribute that must be a number 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "rowspan",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "rowSpan",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 5,
}
`;

exports[`getPropertyInfo gets info for HTML attribute that must be a positive number 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "cols",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "cols",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 6,
}
`;

exports[`getPropertyInfo gets info for HTML boolean attribute 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": true,
  "attributeName": "allowfullscreen",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "allowFullScreen",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 3,
}
`;

exports[`getPropertyInfo gets info for SVG attribute that need special casing 1`] = `null`;

exports[`getPropertyInfo gets info for SVG attribute with the xlink namespace 1`] = `null`;

exports[`getPropertyInfo gets info for SVG attribute with the xml namespace 1`] = `null`;

exports[`getPropertyInfo gets info for attribute that exists both in HTML and SVG 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "tabindex",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "tabIndex",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 1,
}
`;

exports[`getPropertyInfo gets info for attributes that accept URLs 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "xlink:href",
  "attributeNamespace": "http://www.w3.org/1999/xlink",
  "mustUseProperty": false,
  "propertyName": "xlinkHref",
  "removeEmptyString": false,
  "sanitizeURL": true,
  "type": 1,
}
`;

exports[`getPropertyInfo gets info for attributes that accept URLs 2`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "src",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "src",
  "removeEmptyString": true,
  "sanitizeURL": true,
  "type": 1,
}
`;

exports[`getPropertyInfo gets info for different name attribute 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "accept-charset",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "acceptCharset",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 1,
}
`;

exports[`getPropertyInfo gets info for enumerated HTML attribute 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": true,
  "attributeName": "contenteditable",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "contentEditable",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 2,
}
`;

exports[`getPropertyInfo gets info for enumerated SVG attribute 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": true,
  "attributeName": "autoReverse",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "autoReverse",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 2,
}
`;

exports[`getPropertyInfo gets info for reserved prop 1`] = `
PropertyInfoRecord {
  "acceptsBooleans": false,
  "attributeName": "children",
  "attributeNamespace": null,
  "mustUseProperty": false,
  "propertyName": "children",
  "removeEmptyString": false,
  "sanitizeURL": false,
  "type": 0,
}
`;
